aquadoggo/src/materializer/tasks/dependency.rs

```diff
@@ -1,9 +1,293 @@
 // SPDX-License-Identifier: AGPL-3.0-or-later
 
+use p2panda_rs::document::DocumentViewId;
+
 use crate::context::Context;
-use crate::materializer::worker::TaskResult;
+use crate::db::traits::DocumentStore;
+use crate::materializer::worker::{Task, TaskError, TaskResult};
 use crate::materializer::TaskInput;
 
-pub async fn dependency_task(_context: Context, _input: TaskInput) -> TaskResult<TaskInput> {
-    Ok(None)
+/// Helper method for retrieving a document view by it's document view id and if it doesn't exist in
+/// the store, composing a "reduce" task for this specific document view.
+async fn construct_relation_task(
+    context: &Context,
+    document_view_id: DocumentViewId,
+) -> Result<Option<Task<TaskInput>>, TaskError> {
+    match context
+        .store
+        .get_document_view_by_id(&document_view_id)
+        .await
+        .map_err(|_| TaskError::Critical)?
+    {
+        Some(_) => Ok(None),
+        None => Ok(Some(Task::new(
+            "reduce",
+            TaskInput::new(None, Some(document_view_id)),
+        ))),
+    }
+}
+
+pub async fn dependency_task(context: Context, input: TaskInput) -> TaskResult<TaskInput> {
+    // Here we retrive the document view by document view id.
+    let document_view = match input.document_view_id {
+        Some(view_id) => match context
+            .store
+            .get_document_view_by_id(&view_id)
+            .await
+            .map_err(|_| TaskError::Critical)?
+        {
+            Some(document_view) => Ok(document_view),
+            // If no document view for the id passed into this task could be retrieved then this
+            // document has been deleted or the document view id was invalid. As "dependency" tasks
+            // are only dispatched after a successful "reduce" task, neither `None` case should
+            // happen, so this is a critical error.
+            None => Err(TaskError::Critical),
+        },
+        // We only expect to handle document_view_ids in a dependency task.
+        None => Err(TaskError::Critical),
+    }?;
+
+    let mut next_tasks = Vec::new();
+
+    // First we handle all pinned or unpinned relations defined in this tasks document view.
+    // We can think of these as "child" relations.
+    for (_key, document_view_value) in document_view.fields().iter() {
+        match document_view_value.value() {
+            p2panda_rs::operation::OperationValue::Relation(_relation) => {
+                // This is a relation to a document, if it doesn't exist in the db yet, then that
+                // means we either have no entries for this document, or we are not materialising
+                // it for some reason. We don't want to kick of a "reduce" or "dependency" task in
+                // either of these cases, however in the future we may want to flag it as a "wanted"
+                // schema.
```

Comment:
Could be nice to link this to an issue

Reply:
I'm gunna remove the "in the future" comment, it's a little speculative right now...